### PR TITLE
Bugfix acceptances covered by an existing payment that was too low should overdue

### PR DESCRIPTION
**Pull request by Jakub89**
Resolves: #1007 
First commit resolves different bug that I found(someone make sure I'm correct).

### CODE REVIEW

File: concent_api/core/message_handlers.py [hunk 1]
```diff
@@ -820,7 +820,7 @@ def handle_send_force_payment(
 
     # Concent defines time T2 (end time) equal to youngest payment_ts from passed SubtaskResultAccepted messages from
     # subtask_results_accepted_list.
-    payment_ts = min(
+    youngest_payment_ts = max(
```
Comment (cameel) [inline concent_api/core/message_handlers.py]: Yeah, this is a bug and a pretty serious one. It allows the provider to run the settlement algorithm as many times as there are acceptances, getting paid for **all** the acceptances each time.

The fix looks good. If you had a separate pull request, but I can't because I have not looked at the rest of this pull request.

Comment (cameel) [reply concent_api/core/message_handlers.py]: Oh, actually it's not that bad. It's the timestamp in the message so that most it can do is to confuse the provider. I confused it for a moment with `closure_time` in the settlement payment.

Still, it's a bug. Good catch.

Please make sure that `closure_time` in the payment is exactly the same as this `payment_ts`. Actually, I may be a good idea to move the calculation of this timestamp to bankster to have it in one place.

File: concent_api/core/payments/bankster.py [hunk 2]
```diff
@@ -474,3 +477,18 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
         subtask_id=deposit_claim.subtask_id,
     )
     return claim_removed
+
+
+def validate_list_of_transaction_timestamp(
+        list_of_transactions: List[BatchTransferEvent],
```
Comment (rwrzesien) [inline concent_api/core/payments/bankster.py]: Please fix indentation.

File: concent_api/core/message_handlers.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/message_handlers.py]: Please rephrase the comment above by removing `T2` from it, as we no longer use it anywhere in the code

File: concent_api/core/payments/bankster.py [hunk 3]
```diff
@@ -474,3 +477,18 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
...
+        acceptances: List[SubtaskResultsAccepted],
+) -> None:
+    """ Validate if cut_off_time < T2 + PDT, unless requestor already made some payment. """
```
Comment (rwrzesien) [inline concent_api/core/payments/bankster.py]: Same here with the comment.

File: concent_api/core/payments/bankster.py [hunk 4]
```diff
@@ -474,3 +477,18 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
...
+
+    cut_off_time = get_current_utc_timestamp()
+
+    youngest_payments_ts = max(subtask_results_accepted.payment_ts for subtask_results_accepted in acceptances)
+
+    if not any(youngest_payments_ts <= closure_time for closure_time in list_of_transactions):
```
Comment (rwrzesien) [inline concent_api/core/payments/bankster.py]: Those two conditions should be combined into single.

File: concent_api/core/payments/bankster.py [hunk 5]
```diff
@@ -366,6 +367,8 @@ def settle_overdue_acceptances(
             transaction_type=TransactionType.BATCH,
         )
 
+        validate_list_of_transaction_timestamp(list_of_transactions, acceptances)
```
Comment (rwrzesien) [inline concent_api/core/payments/bankster.py]: Split into multiple lines.

File: concent_api/core/tests/test_payments_bankster.py [hunk 6]
```diff
@@ -268,20 +268,22 @@ def test_that_settle_overdue_acceptances_should_return_none_if_subtask_costs_whe
...
-                self._get_list_of_force_transactions(),
-            ]
-        ) as get_list_of_payments_mock:
-            claim_against_requestor = settle_overdue_acceptances(
-                requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
-                provider_ethereum_address=task_to_compute.provider_ethereum_address,
-                acceptances=subtask_results_accepted_list,
-                requestor_public_key=hex_to_bytes_convert(task_to_compute.requestor_public_key),
-            )
+        with mock.patch('core.payments.bankster.validate_list_of_transaction_timestamp', return_value=None) as validate_list_of_transaction_mock:
```
Comment (rwrzesien) [inline concent_api/core/tests/test_payments_bankster.py]: Is there any test that tests this newly created function `validate_list_of_transaction_timestamp`?

File: concent_api/core/message_handlers.py (hunk 1 above)
Comment (Jakub89) [reply concent_api/core/message_handlers.py]: > Please rephrase the comment above by removing `T2` from it, as we no longer use it anywhere in the code.

I did use it also in next commit, where I add function in bankster. I would like to leave it. It can be helpful for someone who know documentation.

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (dybi) [inline concent_api/core/payments/bankster.py]: please reverse the condition, so it's easier to understand
```suggestion
    if any(youngest_payments_ts > closure_time for closure_time in list_of_transactions):
```

File: concent_api/core/tests/test_payments_bankster.py (hunk 6 above)
Comment (dybi) [inline concent_api/core/tests/test_payments_bankster.py]: you doesn't have to provide explicit `return_value` in case you don't expect to get anything out of the function. It is even better not to, because when you set it, it looks as if it was important (an here - it is irrelevant)

File: concent_api/core/tests/test_payments_bankster.py [hunk 7]
```diff
@@ -333,21 +335,23 @@ def test_that_settle_overdue_acceptances_should_return_claim_deposit_with_amount
...
-                        self._get_list_of_force_transactions(),
-                    ]
-                ) as get_list_of_payments_mock:
-                    claim_against_requestor = settle_overdue_acceptances(
-                        requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
-                        provider_ethereum_address=task_to_compute.provider_ethereum_address,
-                        acceptances=subtask_results_accepted_list,
-                        requestor_public_key=hex_to_bytes_convert(task_to_compute.requestor_public_key),
-                    )
+                with mock.patch('core.payments.bankster.validate_list_of_transaction_timestamp', return_value=None) as validate_list_of_transaction_mock:
```
Comment (dybi) [inline concent_api/core/tests/test_payments_bankster.py]: same here

File: concent_api/core/tests/test_payments_bankster.py [hunk 8]
```diff
@@ -424,21 +428,23 @@ def test_that_settle_overdue_acceptances_should_return_claim_deposit_with_amount
...
```
Comment (dybi) [inline concent_api/core/tests/test_payments_bankster.py]: and here

File: concent_api/core/tests/test_payments_bankster.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_bankster.py]: To my surprise there is in `test_integration_force_payment.py`: `test_provider_send_force_payment_beyond_payment_time_concent_should_reject()`

Comment (Jakub89) [reply concent_api/core/tests/test_payments_bankster.py]: There's no corner case being tested where there is `closure_time` > `youngest_payments_ts`, but then it's business as usual :P

File: concent_api/core/payments/bankster.py [hunk 9]
```diff
@@ -474,3 +477,18 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
         subtask_id=deposit_claim.subtask_id,
     )
     return claim_removed
+
+
+def validate_list_of_transaction_timestamp(
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: one more thing - please move it to `validation` module

File: concent_api/core/payments/bankster.py [hunk 10]
```diff
@@ -474,3 +477,18 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
...
+) -> None:
+    """ Validate if cut_off_time < T2 + PDT, unless requestor already made some payment. """
+
+    cut_off_time = get_current_utc_timestamp()
+
+    youngest_payments_ts = max(subtask_results_accepted.payment_ts for subtask_results_accepted in acceptances)
+
+    if not any(youngest_payments_ts <= closure_time for closure_time in list_of_transactions):
+        if cut_off_time < youngest_payments_ts + settings.PAYMENT_DUE_TIME:
+            raise BanksterTimestampError
```
Comment (kbeker) [inline concent_api/core/payments/bankster.py]: TODO from https://github.com/golemfactory/concent/pull/1012. Please add logging in this function

File: concent_api/core/payments/bankster.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: ehh, okey. I will need to change commit name then. So `push --force` here we go! xD

File: concent_api/core/message_handlers.py (hunk 1 above)
Comment (Jakub89) [reply concent_api/core/message_handlers.py]: Done #1020

File: concent_api/core/payments/bankster.py (hunk 2 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done

File: concent_api/core/payments/bankster.py (hunk 3 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: This line change because of [this comment](https://github.com/golemfactory/concent/pull/1015#discussion_r241346283)

Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done

File: concent_api/core/tests/test_payments_bankster.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_bankster.py]: Done

File: concent_api/core/tests/test_payments_bankster.py (hunk 7 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_bankster.py]: Done

File: concent_api/core/tests/test_payments_bankster.py (hunk 8 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_bankster.py]: Done

File: concent_api/core/payments/bankster.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done

File: concent_api/core/payments/bankster.py (hunk 10 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: > They can, but it made more logical sense to me like that, because we don't want to touch the second if, when first isn't True.

This is how Python works, if you have two conditions in `if` connected with `and` and first fails, second is not being run.

File: concent_api/core/tests/test_payments_bankster.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_payments_bankster.py]: @dybi Do you think we should have separate unit test for this function? If not, please resolve this conversation.

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: > I have meant the first condition, should it really raise that exception and stop processing if the list is empty?

I think there's some misunderstanding. First condition doesn't raise any exception. If `list_of_transactions` is empty, we still need to check if `youngest_payments_ts` is not too young. And we raise exception only when it is.

Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: > I think there's some misunderstanding. First condition doesn't raise any exception. If list_of_transactions is empty, we still need to check if youngest_payments_ts is not too young. And we raise exception only when it is.

Makes sense, but it still should be combined into single `if`.

Comment (Jakub89) [reply concent_api/core/payments/bankster.py]: Done
